apps/emqx/test/emqx_passwd_SUITE.erl

```diff
@@ -19,13 +19,70 @@
+    Config.
+
+end_per_suite(_Config) ->
+    ok.
+
+t_hash_data(_) ->
+    Password = <<"password">>,
+    Password = emqx_passwd:hash_data(plain, Password),
+
+    <<"5f4dcc3b5aa765d61d8327deb882cf99">>
+        = emqx_passwd:hash_data(md5, Password),
+
+    <<"5baa61e4c9b93f3f0682250b6cf8331b7ee68fd8">>
+        = emqx_passwd:hash_data(sha, Password),
+
+    <<"5e884898da28047151d0e56f8dc6292773603d0d6aabbdd62a11ef721d1542d8">>
+        = emqx_passwd:hash_data(sha256, Password),
+
+    Sha512 = iolist_to_binary(
+               [<<"b109f3bbbc244eb82441917ed06d618b9008dd09b3befd1b5e07394c706a8bb9">>,
+                <<"80b1d7785e5976ec049b46df5f1326af5a2ea6d103fd07c95385ffab0cacbc86">>]),
+
+    Sha512 = emqx_passwd:hash_data(sha512, Password).
 
 t_hash(_) ->
-    Password = <<"password">>, Salt = <<"salt">>,
-    _ = emqx_passwd:hash(plain, Password),
-    _ = emqx_passwd:hash(md5, Password),
-    _ = emqx_passwd:hash(sha, Password),
-    _ = emqx_passwd:hash(sha256, Password),
-    _ = emqx_passwd:hash(bcrypt, {Salt, Password}),
-    _ = emqx_passwd:hash(pbkdf2, {Salt, Password, sha256, 1000, 20}).
+    Password = <<"password">>,
+    Salt = <<"salt">>,
+    WrongPassword = <<"wrongpass">>,
+
+    Md5 = <<"67a1e09bb1f83f5007dc119c14d663aa">>,
+    Md5 = emqx_passwd:hash({md5, Salt, prefix}, Password),
+    true = emqx_passwd:check_pass({md5, Salt, prefix}, Md5, Password),
+    false = emqx_passwd:check_pass({md5, Salt, prefix}, Md5, WrongPassword),
+
+    Sha = <<"59b3e8d637cf97edbe2384cf59cb7453dfe30789">>,
+    Sha = emqx_passwd:hash({sha, Salt, prefix}, Password),
+    true = emqx_passwd:check_pass({sha, Salt, prefix}, Sha, Password),
+    false = emqx_passwd:check_pass({sha, Salt, prefix}, Sha, WrongPassword),
+
+    Sha256 = <<"7a37b85c8918eac19a9089c0fa5a2ab4dce3f90528dcdeec108b23ddf3607b99">>,
+    Sha256 = emqx_passwd:hash({sha256, Salt, suffix}, Password),
+    true = emqx_passwd:check_pass({sha256, Salt, suffix}, Sha256, Password),
+    false = emqx_passwd:check_pass({sha256, Salt, suffix}, Sha256, WrongPassword),
+
+    Sha512 = iolist_to_binary(
+               [<<"fa6a2185b3e0a9a85ef41ffb67ef3c1fb6f74980f8ebf970e4e72e353ed9537d">>,
+                <<"593083c201dfd6e43e1c8a7aac2bc8dbb119c7dfb7d4b8f131111395bd70e97f">>]),
+    Sha512 = emqx_passwd:hash({sha512, Salt, suffix}, Password),
+    true = emqx_passwd:check_pass({sha512, Salt, suffix}, Sha512, Password),
+    false = emqx_passwd:check_pass({sha512, Salt, suffix}, Sha512, WrongPassword),
+
+    BcryptSalt = <<"$2b$12$wtY3h20mUjjmeaClpqZVve">>,
+    Bcrypt = <<"$2b$12$wtY3h20mUjjmeaClpqZVvehyw7F.V78F3rbK2xDkCzRTMi6pmfUB6">>,
+    Bcrypt = emqx_passwd:hash({bcrypt, BcryptSalt}, Password),
+    true = emqx_passwd:check_pass({bcrypt, BcryptSalt}, Bcrypt, Password),
+    false = emqx_passwd:check_pass({bcrypt, BcryptSalt}, Bcrypt, WrongPassword),
+
+    %% Invalid salt, bcrypt fails
+    <<>> = emqx_passwd:hash({bcrypt, Salt}, Password).
```

Comment:
It's a moment I was thinking a lot. Previous tests didn't verify the result and I was surprised that in some cases  `emqx_passwd:hash` failed.

External users can't provide salt, so they won't be able to bypass the check normally.

The problem is that `bcrypt` does not like salts not generated with `bcrypt:gen_salt`.

What bothers me is the possible situation:

* A new Developer writes a new auth backend.
* They implement salt generation "manually" and then generate hash with `emqx_passwd:hash` (as it was done previously in mnesia backend).
* They make a mistake 🙀 and generate "generic" salt for all hashes, including bcrypt algo.
* So when adding users through API we silently get `<<>>` as hashes.
* When external users come with arbitrary password it is also hashed to `<<>>` because saved salts are invalid for `bcrypt`.
* User bypasses authn.

Currently hash creation is encapsulated in `emqx_authn_password_hashing:hash`, so this is barely possible.

But I'll try to think a bit how to make this completely impossible.